app/models/work_activity.rb

```diff
@@ -40,15 +40,78 @@ def created_by_user
   end
 
   def message_html
-    # convert user references to user links
-    text = message.gsub(USER_REFERENCE) do |at_uid|
-      uid = at_uid[1..-1]
-      user = User.where(uid: uid).first
-      user_info = user&.display_name_safe || uid
-      "<a class='comment-user-link' title='#{user_info}' href='{USER-PATH-PLACEHOLDER}/#{uid}'>#{at_uid}</a>"
-    end
-    # allow ``` for code blocks (Kramdown only supports ~~~)
-    text = text.gsub("```", "~~~")
-    Kramdown::Document.new(text).to_html
+    if activity_type == "CHANGES"
+      changes_html
+    else
+      comment_html
+    end
   end
+
+  private
+
+    def comment_html
+      # convert user references to user links
+      text = message.gsub(USER_REFERENCE) do |at_uid|
+        uid = at_uid[1..-1]
+        user = User.where(uid: uid).first
+        user_info = user&.display_name_safe || uid
+        "<a class='comment-user-link' title='#{user_info}' href='{USER-PATH-PLACEHOLDER}/#{uid}'>#{at_uid}</a>"
+      end
+      # allow ``` for code blocks (Kramdown only supports ~~~)
+      text = text.gsub("```", "~~~")
+      Kramdown::Document.new(text).to_html
+    end
+
+    # Returns the message formatted to display changes that were logged as an activity
+    def changes_html
+      text = ""
+      changes = JSON.parse(message)
+      changes.keys.each do |field|
+        if changes[field].is_a?(Array)
+          # Multi-value change, process each individual entry
+          values = changes[field].map { |value| change_value_html(value) }
+          text += "<p><b>#{field}</b>:#{values.join}</p>"
+        else
+          # Single-value change
+          value = changes[field]
+          text += "<p><b>#{field}</b>: #{change_value_html(value)}</p>"
+        end
+      end
+      text
+    end
+
+    def change_value_html(value)
+      if value["action"] == "added"
+        change_added_html(value["value"])
+      elsif value["action"] == "removed"
+        change_removed_html(value["value"])
+      else
+        change_set_html(value["from"], value["to"])
+      end
+    end
+
+    def change_added_html(value)
+      <<-HTML
+        <i>(added)</i> <span>#{value}</span><br/>
+      HTML
+    end
+
+    def change_removed_html(value)
+      <<-HTML
+        <i>(removed)</i> <span>#{value}</span><br/>
+      HTML
+    end
+
+    def change_set_html(from, to)
+      if from.blank? && to.present?
+        # value was set
+        "<span>#{to}</span><br/>"
+      elsif from.present? && to.present?
+        # value was changed
+        "<span>#{to}</span><br/>"
```

Comment:
Should this be different than line 108?

Reply:
I wanted to highlight that these are different situations (value set for the first time vs value change) in case we want to display that kind of change different to the user. At this point we display it the same, but this makes it easier for somebody to change it if we want something different once users see the log of changes with real data.